core/nightshade/src/nightshade.rs

```diff
@@ -0,0 +1,323 @@
+        let mut result = self.insert_message(hash, message);
+        // Get messages that were blocked by this one.
+        let blocked_messages = match self.missing_hashes.remove(&hash) {
+            None => return result,
+            Some(blocked_messages) => blocked_messages,
+        };
+        for blocked_message_hash in blocked_messages {
+            let is_unblocked = {
+                let (_, unknown_parents) = self.pending_messages.get_mut(&blocked_message_hash).expect("Candidates out of sync");
+                unknown_parents.remove(&hash);
+                unknown_parents.is_empty()
+            };
+            if is_unblocked {
+                let (_, (blocked_message, _)) = self.pending_messages.remove_entry(&blocked_message_hash).expect("Candidates out of sync");
+                let new_result = self.insert_message(blocked_message_hash, blocked_message);
+                result = self.update_result(result, new_result);
+            }
+        }
+        result
+    }
+
+    fn insert_message(&mut self, hash: CryptoHash, message: Message<P>) -> NSResult {
+        let endorses;
+        let mut last_depth = vec![-1; self.num_authorities];
+        let mut max_score = vec![0; self.num_authorities];
+        let mut max_confidence = vec![0; self.num_authorities];
+        if message.parents.is_empty() {
+            endorses = message.author;
+            max_score[message.author] = 1;
+        } else {
+            let mut local_score: Vec<u64> = vec![0; self.num_authorities];
+            for parent in message.parents.iter() {
+                if self.tips.contains(parent) {
+                    self.tips.remove(parent);
+                }
+                let parent_node = self.nodes.get(parent).expect("Checked that parents are present");
+                for i in 0..self.num_authorities {
+                    last_depth[i] = max(last_depth[i], parent_node.last_depth[i]);
+                    max_score[i] = max(max_score[i], parent_node.max_score[i]);
+                    max_confidence[i] = max(max_confidence[i], parent_node.max_confidence[i]);
+                }
+            }
+            for (i, &ld) in last_depth.iter().enumerate() {
+                if ld > self.global_last_depth[i] {
+                    self.global_last_depth[i] = ld;
+                }
+                if ld != -1 {
+                    let node = self.nodes.get(self.nodes_per_author[i].get(&(ld as i64)).expect("Depth should be present")).expect("Node should be present");
+                    if local_score[node.endorses] <= (self.num_authorities * 2 / 3) as u64 {
+                        local_score[node.endorses] += 1;
+                    }
+                }
+            }
+            for i in 0..self.num_authorities {
+                max_score[i] = max(max_score[i], local_score[i]);
+            }
+            let mut candidates: Vec<usize> = (0..self.num_authorities).collect();
```

Comment:
We should just call `min_by` on a vector of tuples (confidence, score, id) by passing our own comparison function.